druid/src/widget/images.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An Image widget.
+//! Please consider using SVG and the SVG wideget as it scales much better.
+
+use std::error::Error;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use image;
+
+use crate::{
+    Affine, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx,
+    PaintCtx, Point, Rect, RenderContext, Size, UpdateCtx, Widget,
+};
+
+use crate::piet::{ImageFormat, InterpolationMode};
+
+#[derive(PartialEq)]
+pub enum FillStrat {
+    Contain,
+    Cover,
+    Fill,
+    FitHeight,
+    FitWidth,
+    None,
+    ScaleDown,
+}
+
+impl Default for FillStrat {
+    fn default() -> Self {
+        FillStrat::Fill
+    }
+}
+
+fn get_scale_offset(parent: Size, fit_box: Size, fit_type: &FillStrat) -> (Point, Point) {
+    let scalex = parent.width / fit_box.width;
+    let scaley = parent.height / fit_box.height;
+
+    let scale: Point = match fit_type {
+        FillStrat::Contain => {
+            let scale = scalex.min(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Cover => {
+            let scale = scalex.max(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Fill => Point {
+            x: scalex,
+            y: scaley,
+        },
+        FillStrat::FitHeight => Point {
+            x: scaley,
+            y: scaley,
+        },
+        FillStrat::FitWidth => Point {
+            x: scalex,
+            y: scalex,
+        },
+        FillStrat::ScaleDown => {
+            let scale = scalex.min(scaley).min(1.0);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::None => Point { x: 1.0, y: 1.0 },
+    };
+
+    let origin_x = (parent.width - (fit_box.width * scale.x)) / 2.0;
+    let origin_y = (parent.height - (fit_box.height * scale.y)) / 2.0;
+    let origin = Point::new(origin_x, origin_y);
+
+    (scale, origin)
+}
+
+/// A widget that renders a Images
```

Comment:
renders an Image ?

Reply:
what would you suggest? `displays`?

Reply:
Is it a typo for `an Image` instead lf `a Images` ? Or we can handle multiple images ?

Reply:
Oh i see, thanks for the spot